codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/CodegenDelegator.kt

```diff
@@ -114,3 +116,24 @@ fun CodegenWriterDelegator<RustWriter>.finalize(
     }
     flushWriters()
 }
+
+internal fun mergeDependencyFeatures(cargoDependencies: List<CargoDependency>): List<CargoDependency> {
```

Comment:
I think this looks correct, but a potential refactoring to simplify it a bit:
1. Add `.key()` to `CargoDependency` that returns the tuple of `name, location, scope`.
2.
```kotlin
fun CargoDependency.mergeWith(other: CargoDependency): CargoDependency {
  check(this.key() == other.key())
  this.copy(features = ..., option = ...)
}
internal fun mergeDependencyFeatures(dependencies: List<CargoDependency>): List<CargoDependency> = 
  dependencies
    .groupBy { it.key() }
   .mapValues { group -> group.reduce { acc, next -> acc.mergeWith(next) } }
```